Pica improvements

bunnei opened the pull request:
This is a combination of several Pica emulation improvements. All changes have been implemented for both the SW rasterizer and OpenGL rasterizer. I recommend reviewing each commit separately, as they are well separated. This PR implements:
* Pica command buffer execution registers
* Vertex shader MIN, SLT, and SLTI instructions
* Color "logic operation" (same as `glLogicOp`)
* ~~RG8 texture format~~
* AddSigned combiner function for the alpha channel

Here are some results of these features (although I'm sure there are many more, testing welcome!):
![](http://i.imgur.com/RCHE31L.png)![](http://i.imgur.com/AQ8RMYo.png)

Subv commented on src/video_core/vertex_shader.cpp:
```diff
@@ -119,17 +119,13 @@ static void ProcessShaderCode(VertexShaderState& state) {
 
-            const float24* src1_ = LookupSourceRegister(instr.common.GetSrc1(is_inverted) + address_offset);
-            const float24* src2_ = LookupSourceRegister(instr.common.GetSrc2(is_inverted));
+            const float24* src1_ = LookupSourceRegister(instr.common.GetSrc1(is_inverted) + (!is_inverted ? address_offset : 0));
```
Is this really all that was needed to implement the inverted instruction format?

neobrain replied:
Suggestion: => `!is_inverted * address_offset`

neobrain replied:
@Subv yup.

neobrain commented on src/video_core/color.h:
```diff
@@ -70,6 +70,17 @@ inline const Math::Vec4<u8> DecodeRGB8(const u8* bytes) {
+ * @param bytes Pointer to encoded source color
+ * @return Result color decoded as Math::Vec4<u8>
+ */
+inline const Math::Vec4<u8> DecodeRG8(const u8* bytes) {
```
Sure that this is RG, rather than GR?

ILOVEPIE replied:
Did you really just copy my #773 pull request? that was closed for lack of having a good test case yet so this shouldn't be merged until we have a test case.

neobrain commented on src/video_core/pica.h:
```diff
@@ -707,8 +723,27 @@ struct Regs {
+        BitField< 0, 20, u32> size_1;
+        BitField< 0, 28, u32> addr_0;
+        BitField< 0, 28, u32> addr_1;
+        u32 trigger_0;
```
Why not make this an array?

neobrain replied:
*these, I guess.

neobrain commented on src/video_core/pica.h:
```diff
@@ -707,8 +723,27 @@ struct Regs {
+        PAddr GetPhysicalAddress(int index) {
+            return (PAddr)(8 * ((index == 0) ? addr_0 : addr_1));
+        }
+    } cmd_buffer;
```
No abbreviations plz. => `command_buffer`

neobrain commented on src/video_core/rasterizer.cpp:
```diff
@@ -402,9 +402,11 @@ static void ProcessTriangleInternal(const VertexShader::OutputVertex& v0,
+
+                    // HACK: Until we implement fragment lighting, just use primary_color
                     case Source::PrimaryFragmentColor:
+                    case Source::SecondaryFragmentColor:
```
Return 0 instead for this branch. primary_color makes no sense at all.

neobrain commented on src/video_core/rasterizer.cpp:
```diff
@@ -570,6 +572,10 @@ static void ProcessTriangleInternal(const VertexShader::OutputVertex& v0,
 
+                    case Operation::AddSigned:
+                        // TODO(bunnei): Verify that the color conversion from (float) 0.5f to (byte) 128 is correct
+                        return MathUtil::Clamp<u8>(int(input[0]) + int(input[1]) - 128, 0, 255);
```
is the int cast actually necessary there? if it is, use static_cast instead.

Also note that this clamp is buggy: casting the `Clamp` inputs to u8 will over-/underflow in all cases where the Clamp would have any effect.

neobrain replied:
Actually, given that this was submitted in a broken way without the author noticing it (suggesting a lack of proper test case), I'd suggest dropping it from the change set until a hardware test is written.

ILOVEPIE replied:
A good test for this is Animal Crossing.

bunnei replied:
@neobrain I'm going to leave it in. Mistakes happen, that's the point of the PR process. It's fixed and is correct now, with the except of the 0.5f -> u8 conversion, which has a TODO to verify (as it does in other places in the code).

neobrain commented on src/video_core/rasterizer.cpp:
```diff
@@ -866,8 +871,69 @@ static void ProcessTriangleInternal(const VertexShader::OutputVertex& v0,
-                UNIMPLEMENTED();
+                static auto LogicOp = [](const Math::Vec4<u8>& src, const Math::Vec4<u8>& dest, Regs::LogicOp op) -> Math::Vec4 < u8 > {
+                    switch (op) {
+                    case Regs::LogicOp::Clear:
```
Don't submit untested code. Just implement what you actually found used, and of which you asserted that it works in a more sophisticated way than "I didn't see any glitches on the screen". Alternatively, write an automatable hardware test for it.

bunnei replied:
It's tested

neobrain commented on src/video_core/rasterizer.cpp:
```diff
@@ -866,8 +871,69 @@ static void ProcessTriangleInternal(const VertexShader::OutputVertex& v0,
+                                                 src.b() | ~dest.b(), src.a() | ~dest.a());
+
+                    case Regs::LogicOp::OrInverted:
+                        return Math::MakeVec<u8>(~src.r() | dest.r(), ~src.g() | dest.g(),
```
What's the result of bitwise operators on `uint8_t`? If it's u8 again, there's no need to explicitly specify the template parameter for MakeVec.

ILOVEPIE replied:
result is int

bunnei replied:
It's `int`, so they are indeed necessary

lioncash commented on src/video_core/pica.h:
```diff
@@ -707,8 +723,27 @@ struct Regs {
+            return 8 * ((index == 0) ? size_0 : size_1);
+        }
+
+        PAddr GetPhysicalAddress(int index) {
```
const function.

neobrain replied:
unsigned

lioncash commented on src/video_core/pica.h:
```diff
@@ -707,8 +723,27 @@ struct Regs {
+        u32 trigger_0;
+        u32 trigger_1;
+
+        int GetSize(int index) {
```
const function

neobrain replied:
Also the types involved here should all be unsigned.

neobrain commented on src/video_core/vertex_shader.cpp:
```diff
@@ -119,17 +119,13 @@ static void ProcessShaderCode(VertexShaderState& state) {
-            //       offset needs to be applied to SRC2 instead, etc.
-            //       For now, we just abort in this situation.
-            ASSERT_MSG(!is_inverted, "Bad condition...");
+            const bool is_inverted = 0 != (instr.opcode.Value().GetInfo().subtype & OpCode::Info::SrcInversed);
```
Please put parentheses around comparisions when their result is being assigned to a variable.

neobrain commented on src/video_core/pica.h:
```diff
@@ -707,8 +723,27 @@ struct Regs {
+    INSERT_PADDING_WORDS(0x2);
+
+    struct {
+        BitField< 0, 20, u32> size_0;
```
should have documentation wrt units.

neobrain commented on src/video_core/command_processor.cpp:
```diff
@@ -58,6 +58,16 @@ static inline void WritePicaReg(u32 id, u32 value, u32 mask) {
+            u32* head_ptr = (u32*)Memory::GetPhysicalPointer(g_state.regs.cmd_buffer.GetPhysicalAddress(index));
+            g_state.cmd_list.head_ptr = g_state.cmd_list.current_ptr = head_ptr;
+            g_state.cmd_list.length = g_state.regs.cmd_buffer.GetSize(index) / sizeof(u32);
+            return;
```
Why does this "return" instead of "break" (same for the case above, actually) ?

neobrain replied:
If this has an actual reason, please add a comment about it.

SeannyM commented:
Tested these changes, the intro of Phoenix Wright Dual Destinies is working better now, this effect previously didn't work:
![capcom](https://cloud.githubusercontent.com/assets/3057954/7803139/1b3a9e52-0316-11e5-8349-15ff66e56fb1.png)
...but I'm getting a black screen from Ocarina of Time on the OpenGL renderer. It still works on software.

bunnei commented:
Addressed comments. There was a regression that broke some games (pointed out by @SeannyM) that should be fixed now. I also removed the RG8 texture format decoding; while Zelda: A Link Between Worlds used this format, it wasn't using it in a way that was trivial to verify correctness.

SeannyM commented:
@bunnei everything working as it should now on my end.

JohnGodgames commented:
Some of regression:
-White color instead of textures which worked before...


Harvest Moon: A New Beginning
![harvest moon](https://cloud.githubusercontent.com/assets/11453507/7805998/b99e156e-037b-11e5-8782-42fed34321ca.PNG)


Pokémon Rumble Blast
![pokemon](https://cloud.githubusercontent.com/assets/11453507/7806003/d055fce0-037b-11e5-9fb2-710e40074c96.PNG)

Subv commented on src/video_core/command_processor.cpp:
```diff
@@ -363,38 +373,34 @@ static inline void WritePicaReg(u32 id, u32 value, u32 mask) {
-    read_pointer += 2;
+    while (g_state.cmd_list.current_ptr < g_state.cmd_list.head_ptr + g_state.cmd_list.length) {
+        const u32* first_command_word = g_state.cmd_list.current_ptr;
+        const CommandHeader header = *(const CommandHeader*)&first_command_word[1];
```
static_cast please

yuriks replied:
I'd prefer this instead: `const CommandHeader header = { first_command_word[1] };`

yuriks replied:
Actually, `CommandHeader` has a constructor, so just `const CommandHeader header = first_command_word[1];` will already work.

neobrain replied:
There's no reason to have a constructor in CommandHeader. Instead, the constructor should be removed and `const CommandHeader header = { first_command_word[1] };` be used instead.

yuriks replied:
I looked into this and it turns out that for some reason we have *two* definitions of `CommandHeader` in the codebase, one in `video_core/pica.h` and another in `command_processor.h`. The difference is one is documented and the other has a constructor defined. One of those should be removed.

Subv commented:
Did you test the logic ops? if so LGTM.

yuriks commented on src/video_core/command_processor.cpp:
```diff
@@ -363,38 +373,34 @@ static inline void WritePicaReg(u32 id, u32 value, u32 mask) {
-    return read_pointer - first_command_word;
-}
+        for (unsigned i = 0; i < 1 + header.extra_data_length; ++i) {
+            ++g_state.cmd_list.current_ptr;
```
`+= 1` instead? I always find standalone pre/post increments a bit off-putting.

neobrain replied:
Sounds like a matter of personal taste. We do standalone increments all over the place, and IMO they're perfectly fine :/

yuriks commented on src/video_core/command_processor.cpp:
```diff
@@ -363,38 +373,34 @@ static inline void WritePicaReg(u32 id, u32 value, u32 mask) {
+        }
+        
+        // align read pointer to 8 bytes
+        if ((first_command_word - g_state.cmd_list.current_ptr) % 2)
```
Shouldn't this use `head_ptr` instead of `first_command_word` to prevent a bogus alignment on a buffer change? (This will never cause problems since both buffers will most likely have 16-byte alignment in the first place, but is still sloppy.)

yuriks commented on src/video_core/command_processor.cpp:
```diff
@@ -363,38 +373,34 @@ static inline void WritePicaReg(u32 id, u32 value, u32 mask) {
-    while (read_pointer < list + list_length) {
-        read_pointer += ExecuteCommandBlock(read_pointer);
+            cmd = header.cmd_id + ((header.group_commands) ? i + 1 : 0);
+            value = *g_state.cmd_list.current_ptr;
```
This part is pretty confusing, imo. I tried to make it a bit more clearer: https://gist.github.com/yuriks/84a676187ebbe5d6c06e

yuriks commented on src/video_core/command_processor.cpp:
```diff
@@ -363,38 +373,34 @@ static inline void WritePicaReg(u32 id, u32 value, u32 mask) {
+        const u32 write_mask = ((header.parameter_mask & 0x1) ? (0xFFu << 0) : 0u) |
+            ((header.parameter_mask & 0x2) ? (0xFFu << 8) : 0u) |
+            ((header.parameter_mask & 0x4) ? (0xFFu << 16) : 0u) |
+            ((header.parameter_mask & 0x8) ? (0xFFu << 24) : 0u);
```
http://ideone.com/hcFibk *hint* *hint* *nudge* *nudge*

yuriks commented on src/video_core/rasterizer.cpp:
```diff
@@ -570,6 +575,13 @@ static void ProcessTriangleInternal(const VertexShader::OutputVertex& v0,
+                    {
+                        // TODO(bunnei): Verify that the color conversion from (float) 0.5f to (byte) 128 is correct
+                        auto result = static_cast<int>(input[0]) + static_cast<int>(input[1]) - 128;
+                        return static_cast<u8>(MathUtil::Clamp<int>(result, 0, 255));
```
Is this static_cast necessay?

yuriks commented on src/video_core/rasterizer.cpp:
```diff
@@ -866,8 +877,63 @@ static void ProcessTriangleInternal(const VertexShader::OutputVertex& v0,
+                    LogicOp(combiner_output.r(), dest.r(), output_merger.logic_op),
+                    LogicOp(combiner_output.g(), dest.g(), output_merger.logic_op),
+                    LogicOp(combiner_output.b(), dest.b(), output_merger.logic_op),
+                    LogicOp(combiner_output.a(), dest.a(), output_merger.logic_op));
```
You can make `LogicOp` operate on a `u32` and do all of these in one go.

neobrain replied:
... but please don't actually do that. There's little point and it just makes the code less obvious to understand IMO.

yuriks commented on src/video_core/renderer_opengl/gl_state.cpp:
```diff
@@ -99,8 +101,11 @@ void OpenGLState::Apply() {
         if (blend.enabled) {
             glEnable(GL_BLEND);
+            glDisable(GL_COLOR_LOGIC_OP);
+            glLogicOp(GL_COPY);
```
This should also set the corresponding flags in the state so that they don't go out of sync.

yuriks commented:
Pass done.

yuriks commented:
You might also want to look at @JohnGodgames's regressions and see if they're simply caused by the stubbed implementation of fragment lighting.

tfarley commented on src/video_core/renderer_opengl/gl_state.h:
```diff
@@ -42,6 +42,8 @@ class OpenGLState {
         } color; // GL_BLEND_COLOR
     } blend;
 
+    GLenum logic_op; // GL_COPY
```
This comment should be "// GL_LOGIC_OP_MODE" to correspond to its glGet*() like the others (since the GL_COPY default is already specified in constructor).

yuriks commented:
LGTM

JohnGodgames commented:
Just tested an update and there is a result:

pros:

Fixed
![pica pokemon rb](https://cloud.githubusercontent.com/assets/11453507/7858982/1f43e084-053d-11e5-96de-30ad23720dae.PNG)

Fixed
![pica harvest moon](https://cloud.githubusercontent.com/assets/11453507/7859098/0f953ccc-053e-11e5-86cc-b39e347095d6.PNG)


cons:

Stiil not perfect
![pica sonic](https://cloud.githubusercontent.com/assets/11453507/7858999/4672565e-053d-11e5-9212-5ed5c96e0c25.PNG)

neobrain commented on src/video_core/command_processor.cpp:
```diff
@@ -363,38 +373,34 @@ static inline void WritePicaReg(u32 id, u32 value, u32 mask) {
-
-    return read_pointer - first_command_word;
+/// Expands a 4-bit mask to 4-byte mask, e.g. 0b0101 -> 0x00FF00FF
+static u32 ExpandBitsToBytes(u32 x) {
```
Wouldn't this be easier to grasp if we used the old code (i.e. using ternaries) in this function instead of this gibberish? The old code was obviously correct, while this new one is pretty confusing (I couldn't verify its correctness without pencil+paper ...)

yuriks replied:
You can easily verify it's correctness just by executing the function for all the 16 possible inputs. (Which is what I did.)

The function is documented on what it does, and this code is never going to change, so the ability for it to be easily modified isn't important.

neobrain replied:
easy != obvious.

I'm not seeing how this voodoo (no offense) is better than plain obvious code.

hrydgard replied:
This one is probably faster to do with a lookup-table anyway actually, as it's used quite a bit so the tiny 16-entry table will sit in L1 or L0

yuriks replied:
Good point about the table.

neobrain replied:
@hrydgard I don't think performance considerations matter here (since this is far from being a hotspot), but I guess a LUT-based approach would be both performant and readable ;)

neobrain commented on src/video_core/command_processor.cpp:
```diff
@@ -363,38 +373,34 @@ static inline void WritePicaReg(u32 id, u32 value, u32 mask) {
+        if ((g_state.cmd_list.head_ptr - g_state.cmd_list.current_ptr) % 2 != 0)
+            ++g_state.cmd_list.current_ptr;
+
+        const CommandHeader header = { g_state.cmd_list.current_ptr[1] };
```
This can be reordered and rewritten to:

    u32 value = *g_state.cmd_list.current_ptr++;
    const CommandHeader header = { *g_state.cmd_list.current_ptr++ };

which doesn't require a standalone "g_state.cmd_list.current_ptr += 2;" line and in general is easier to read IMO.

hrydgard replied:
Either way you really should move current_ptr into a local variable before the start of the loop, and put it back afterwards. Walking pointers stored in global variables is actually really slow.

yuriks replied:
current_ptr is modified inside WritePicaReg, unfortunately, so it has to be read at least once per loop iteration.

neobrain commented on src/video_core/pica.h:
```diff
@@ -707,8 +723,31 @@ struct Regs {
+
+    struct {
+        // There are two channels that can be used to configure the next command buffer, which
+        // can be then executed (without a CPU interrupt) by writing to the "trigger" registers
```
The "regular" method to submit GPU commands via GSP doesn't require CPU interrupts either, or does it? Seems like it's not really made clear what the actual point here is.

bunnei replied:
@neobrain yes a normal command list from GSP always ends with an interrupt

neobrain replied:
@bunnei that doesn't have anything to do with launching other command lists, though...

bunnei replied:
I still think it's relevant information, because that's why these registers are used :)

neobrain commented on src/video_core/pica.h:
```diff
@@ -707,8 +723,31 @@ struct Regs {
+        // There are two channels that can be used to configure the next command buffer, which
+        // can be then executed (without a CPU interrupt) by writing to the "trigger" registers
+        // below. An arbitrary number of additional command buffers may be executed in a row. Games
+        // often configure these registers to provide a command list "jump" mechanism.
```
"Games often configure these registers to provide a command list "jump" mechanism." ... isn't this exactly the point? This phrasing suggests that it could be configured in a way which doesn't "jump".

bunnei replied:
What I was trying to say is not clear, by "jump" I mean "subroutine" (as in, jump to a command list, then return). I'll fix the comment

neobrain replied:
yeah, subroutine is a lot clearer indeed.

neobrain commented on src/video_core/vertex_shader.cpp:
```diff
@@ -279,6 +284,18 @@ static void ProcessShaderCode(VertexShaderState& state) {
 
+            case OpCode::Id::SLT:
+            case OpCode::Id::SLTI:
+            {
```
redundant braces

neobrain commented on src/video_core/pica.h:
```diff
@@ -707,8 +723,34 @@ struct Regs {
+        // There are two channels that can be used to configure the next command buffer, which
+        // can be then executed by writing to the "trigger" registers. There are two reasons why a
+        // game might use this feature:
+        //  1) With this, an arbitrary number of additional command buffers may be executed without
```
suggested alternative: "An arbitrary number of additional command buffers may be queued without explicitly having the CPI wait for completion of the current command buffer execution"

yuriks replied:
Meh, I think the current wording for the first part is more accurate, though it meanders at the end.

How about this?
> With this, an arbitrary number of additional command buffers may be executed in sequence without requiring intervention of the CPU after the initial one is kicked off.
